lib/provisioning/Provisioner.js

```diff
@@ -91,62 +159,195 @@ Provisioner.prototype.isProvisionRequest = function(req) {
+                            `Known ops in this channel: ${knownOps}`);
+        }
+
+        // (Matrix) update room state
+        log.info(`Sending m.room.bridging`);
+        // State key for m.room.bridging
+        let skey = `irc://${ircDomain}/${ircChannel}`;
+
+        // Send pending m.room.bridging
+        yield this._updateBridgingState(roomId, userId, 'pending', skey);
 
-Provisioner.prototype._validate = function(actual, parameterName) {
-    let valid = parameterValidation[parameterName];
+        // (IRC) Ask operator for authorisation
+        // Time that operator has to respond before giving up
+        let timeoutSeconds = this._requestTimeoutSeconds;
 
-    if (!valid) {
-        throw new Error(
-            `Parameter name not recognised (${parameterName}).`
-        );
+        // Deliberately not yielding on this so that 200 OK is returned
+        log.info(`Contacting operator`);
+        this._contactOperator(
+            botClient, server, opNick, ircChannel,
+            roomId, userId, skey, timeoutSeconds);
     }
+);
+
+Provisioner.prototype._contactOperator = Promise.coroutine(
+    function*(botClient, server, opNick, ircChannel, roomId, userId, skey, timeoutSeconds) {
+        let d = promiseutil.defer();
+
+        // Send PM to operator
+        let authRequestAction = new IrcAction("message",
+            `${userId} has requested to bridge ${roomId} with ${ircChannel} on this IRC ` +
+            `network. Respond with 'yes' or 'y' to allow, or simply ignore this message to ` +
+            `dissallow. You have ${timeoutSeconds} seconds from when this message was sent.`);
+        let pmRoom = new IrcRoom(server, opNick);
 
-    if (!actual) {
-        throw new Error(
-            `${parameterName} not provided (like '${valid.example}').`
-        );
+        this._setRequest(server, opNick, {userId: userId, defer: d});
+
+        this._ircBridge.sendIrcAction(pmRoom, botClient, authRequestAction);
+        try {
+            yield d.promise.timeout(timeoutSeconds * 1000);
+            this._removeRequest(server, opNick);
+        }
+        catch (err) {
+            log.info(`Operator ${opNick} did not respond (${err.message})`);
+            yield this._updateBridgingState(roomId, userId, 'failure', skey);
+            this._removeRequest(server, opNick);
+            return;
+        }
+        try {
+            yield this._doLink(server, ircChannel, roomId);
+        }
+        catch (err) {
+            log.info(`Failed to create link following authorisation (${err.message})`);
+            yield this._updateBridgingState(roomId, userId, 'failure', skey);
+            this._removeRequest(server, opNick);
+            return;
+        }
+        yield this._updateBridgingState(roomId, userId, 'success', skey);
     }
+);
+
+Provisioner.prototype._removeRequest = function (server, opNick) {
+    delete this._pendingRequests[server.domain][opNick];
```

Comment:
But deleting something that isn't there does nothing, which is expected in this case?